lib/SILOptimizer/Mandatory/TFPartition.cpp

```diff
@@ -942,7 +972,10 @@ diagnoseCopyToAccelerator(SILValue value, SILInstruction *user,
 
   auto &ctx = hostFn.getModule().getASTContext();
 
-  // Emit the warning.
+  // Emit the warning on this value, if that has not been done before.
+  if (!copyToAccelWarningLocs.insert(loc.getSourceRange()).second)
+    return;
+
   diagnose(ctx, loc.getSourceLoc(), diagID, description)
```

Comment:
Do you need to unique based on the entire source range?  Would it be fine to just unique based on loc.getSourceLoc() ?   That would eliminate the need for template specialization entirely.

Reply:
`DenseMapInfo` is not defined for `SourceLoc` either, so I might as make `SourceRange` work as the key type.

Compiler error for using `llvm::DenseSet<SourceLoc> copyToHostWarningLocs`:
```
/usr/local/google/home/hongm/ssd_part/git/swift-base/llvm/include/llvm/ADT/DenseMap.h:421:22: error: no member named 'getEmptyKey' in 'llvm::DenseMapInfo<swift::SourceLoc>'
    return KeyInfoT::getEmptyKey();
```